### PR TITLE
Tickets/dm 31537

### PR DESCRIPTION
**Pull request by jgates108**

### CODE REVIEW

File: src/wsched/ScanScheduler.cc [hunk 1]
```diff
@@ -221,23 +209,59 @@ util::Command::Ptr ScanScheduler::getCmd(bool wait)  {
...
+    bool first = true;
+    QueryId qid;
+    int jid = 0;
+    // Convert to a vector of tasks
+    for (auto const& cmd:cmds) {
+        wbase::Task::Ptr t = dynamic_pointer_cast<wbase::Task>(cmd);
+        if (t == nullptr) {
+            LOGS(_log, LOG_LVL_ERROR, getName()
+                    << " queCmd could not be converted to Task or was nullptr");
+            return;
```
Comment (andy-slac) [inline src/wsched/ScanScheduler.cc:230]: Why return, there are other tasks to process? If this is a condition that should never happen then throwing an exception is probably better?

**Review by andy-slac — APPROVED** [review]
Looks ok in general, though my understanding of the code is not great. Couple of general concerns:
- Removing `std::` does not improve the code but adds a lot of noise and strain to my eyes. There is no good reason to do that for exiting code. 
- I think that code complexity continues to increase, we may want to try to redesign and refactor parts of qserv code before it becomes too complex to understand.

File: src/ccontrol/MergingHandler.h [hunk 2]
```diff
@@ -115,8 +116,9 @@ class MergingHandler : public qdisp::ResponseHandler {
     MsgState _state; ///< Received message state
     std::shared_ptr<proto::WorkerResponse> _response; ///< protobufs msg buf
     bool _flushed {false}; ///< flushed to InfileMerger?
-    std::string _wName {"~"}; /// worker name
-    std::mutex _setResultMtx; // Allow only one call to ParseFromArray at a time from _seResult.
+    std::string _wName {"~"}; ///< worker name
+    std::mutex _setResultMtx; //< Allow only one call to ParseFromArray at a time from _seResult.
+    std::set<int> _jobIds; ///< Set of jobIds added in this request.
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/ccontrol/MergingHandler.h:121]: I went with set to make certain there were no duplicate entries.

File: src/proto/worker.proto [hunk 3]
```diff
@@ -90,14 +91,15 @@ message ProtoHeader {
...
 }
 
 message ColumnSchema {
     optional string name = 1; // Optional to allow type-only transmission
-    required bool deprecated_hasdefault = 2; // DEPRECATED (but 'required', so will have to continue to be set)
-    optional bytes deprecated_defaultvalue = 3; // DEPRECATED
-    required string sqltype = 4;
-    optional int32 mysqltype = 5;
+    required string sqltype = 2;
+    optional int32 mysqltype = 3;
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/proto/worker.proto:102]: At this point, with the change from header separate from data to header appended to previous data, backwards compatability is not possible. Seems like a good time for cleanup.

File: src/wdb/QueryRunner.cc [hunk 4]
```diff
@@ -466,93 +387,177 @@ class ChunkResourceRequest {
...
+        }
+        tSize += rawRow->ByteSizeLong();
+        ++rowCount;
+
+        unsigned int szLimit = std::min(proto::ProtoHeaderWrap::PROTOBUFFER_DESIRED_LIMIT,
+                                        proto::ProtoHeaderWrap::PROTOBUFFER_HARD_LIMIT);
+
+        // This thread may have already been removed from the pool for
+        // other reasons, such as taking too long.
+        if (not _removedFromThreadPool) {
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/wdb/QueryRunner.cc:420]: There are other cases where a thread can be removed from the pool. I don't think this check is any more expensive than the check for first time through the loop would be.

File: src/proto/worker.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/proto/worker.proto:102]: According to @fritzm, any changes in the Protobuf attributes require changing the numbers. This will prevent mysterious crashes in case if (intentionally or accidentally) different versions of the Protobuf schema will be deployed at a client and a server. I think what's done here was the right thing to do.

File: src/wdb/QueryRunner.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (jgates108) [reply src/wdb/QueryRunner.cc:420]: It may be possible to move this out of the loop. The code here has changed significantly, so I'll try that.

File: src/proto/worker.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (fritzm) [reply src/proto/worker.proto:102]: There is a difference between having mismatched versions accidentally running at the same time causing clean error fails, and having mismatched versions accidentally running at the same time causing exploding segvs of death...

Also, the point is not that numbers have to _change_, but that they should _never be recycled_...

[the remark this answers is not in the document]
Comment (jgates108) [reply src/proto/worker.proto:102]: I think that's a change on arguments I've heard in the past, where there was some hope of backward compatibility. There's no hope of backward compatibility in this change. Hoping for backwards compatibility seems risky to me.

The new argument seems to be to never reuse numbers, so it fails loudly an obviously when there's a protocol issue between worker/czar. That seems safer.

There is a small benefit to using small numbers as those <=15 only use one byte to encode, so the messages are smaller. This may actually matter when dealing with result rows.

File: src/qdisp/Executive.cc [hunk 5]
```diff
@@ -410,7 +418,14 @@ void Executive::_setup() {
...
     } else {
-        _xrdSsiService = XrdSsiProviderClient->GetService(eInfo, _config.serviceUrl.c_str()); // Step 1
+        if (_xrdSsiServiceStatic == nullptr) {
+            // In operations, this would be locked once per program run.
+            lock_guard<mutex> xrdSsiServLock(xrdSsiServiceMtx);
+            if (_xrdSsiServiceStatic == nullptr) {
+                _xrdSsiServiceStatic = XrdSsiProviderClient->GetService(eInfo, _config.serviceUrl.c_str());
+            }
+        }
+        _xrdSsiService = _xrdSsiServiceStatic; // Step 1
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/qdisp/Executive.cc:428]: Daniel had the steps of the XrdSsi connection commented in the code. I haven't really felt the need to remove them.

File: src/qdisp/QueryRequest.cc [hunk 6]
```diff
@@ -407,74 +448,103 @@ void QueryRequest::ProcessResponseData(XrdSsiErrInfo const& eInfo,
...
-                                                              largeResult, nextBufSize, resultRows);
+    bool last = false;
+    bool flushOk = false;
+
+    // The buffer has 2 parts.
+    // - The first (bytes = blen - ProtoHeaderWrap::getProtheaderSize())
+    //   is the result associated with the previously received header.
+    // - The second is the header for the next message.
+    int respSize = blen - protoHeaderSize;
+    nextHeaderBufPtr = make_shared<vector<char>>(bufPtr->begin() + respSize, bufPtr->end());
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/qdisp/QueryRequest.cc:478]: The header is 256 bytes or so, the rest of the buffer is potentially several megabytes. I chose to copy the small part so I could delete the large part as soon as it wasn't needed. I think it also resulted in simpler code.

File: src/qdisp/QueryRequest.cc [hunk 7]
```diff
@@ -407,74 +448,103 @@ void QueryRequest::ProcessResponseData(XrdSsiErrInfo const& eInfo,
...
     if (largeResult) {
         if (!_largeResult) LOGS(_log, LOG_LVL_DEBUG, "holdState largeResult set to true");
         _largeResult = true; // Once the worker indicates it's a large result, it stays that way.
     }
 
     if (flushOk) {
-        _totalRows += resultRows;
+        if (last != xrdLast) {
+            LOGS(_log, LOG_LVL_DEBUG, "processData disagreement between last=" << last
+                                     << " and xrdLast=" << xrdLast);
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/qdisp/QueryRequest.cc:512]: xrdLast is inconsistent, while last is authoratative/correct. This is just a footnote in case it becomes important to figure out why xrootd disagrees on last. Not really a happy fun thing, but it has been around for many years.

File: src/qdisp/QueryRequest.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (andy-slac) [reply src/qdisp/QueryRequest.cc:478]: Reducing number of allocations is a good idea in general, and allocating buffers of very different sizes from the same memory arena can cause memory fragmentation.

File: src/proto/ProtoHeaderWrap.cc [hunk 8]
```diff
@@ -59,7 +63,7 @@ std::string ProtoHeaderWrap::wrap(std::string& protoHeaderString) {
 }
 
 bool ProtoHeaderWrap::unwrap(std::shared_ptr<WorkerResponse>& response, std::vector<char>& buffer) {
-    response->headerSize = static_cast<unsigned char>(buffer[0]);
+    response->headerSize = static_cast<unsigned char const>(buffer[0]);
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/proto/ProtoHeaderWrap.cc:66]: Does it hurt? Changing the contents of the buffer would be bad.

[the remark this answers is not in the document]
Comment (andy-slac) [reply src/proto/ProtoHeaderWrap.cc:66]: It does not do anything, and it's not preventing any potential changes to buffer (because there could not be any changes in this statement). It's sort of thing like this:

    int a = 1;
    int b = static_cast<int const>(a);

which is legal but pointless.

File: src/qdisp/QueryRequest.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (jgates108) [reply src/qdisp/QueryRequest.cc:478]: It can be a pretty long time (possibly many minutes) before the need for the nextHeaderBuffer goes away. Considering that there can be thousands of these, I think freeing the large block of memory as early as possible is more important than possible memory fragmentation.

File: src/util/InstanceCount.cc [hunk 9]
```diff
@@ -46,7 +46,7 @@ void InstanceCount::_increment(std::string const& source) {
     auto ret = _instances.insert(entry);
     auto iter = ret.first;
     iter->second += 1;
-    LOGS(_log, LOG_LVL_DEBUG, "InstanceCount " << source
+    LOGS(_log, LOG_LVL_INFO, "InstanceCount " << source
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/util/InstanceCount.cc:49]: There should only be one instance of InstanceCount in the code in GetBuff(), and there should only be on of those at any given time, so it should be pretty quiet. This is to help locate the count( * ) hanging Task proble, which is proving to be difficult. If there's a hanging GetBuff() call, that's hard to figure out without this in the log. The alternate method is gdb on the worker in the docker container, dump thread info to a logfile, close gdb, search for GetBuff in the log file.

File: src/util/InstanceCount.cc [hunk 10]
```diff
@@ -56,7 +56,7 @@ InstanceCount::~InstanceCount() {
     auto iter = _instances.find(_className);
     if (iter != _instances.end()) {
         iter->second -= 1;
-        LOGS(_log, LOG_LVL_DEBUG, "~InstanceCount " << iter->first << "=" << iter->second << " : " << *this);
+        LOGS(_log, LOG_LVL_INFO, "~InstanceCount " << iter->first << "=" << iter->second << " : " << *this);
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/util/InstanceCount.cc:59]: Same answer.

File: src/wbase/SendChannelShared.h [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+namespace qserv {
+
+namespace wcontrol {
+class TransmitMgr;
+class TransmitLock;
+}
+
+namespace wbase {
+
+/// A class that provides a SendChannel object with synchronization so it can be
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/wbase/SendChannelShared.h:49]: The SendChannelShared wraps a SendChannel with code needed for synchronization. Igor's code uses a lot of SendChannel instances and the addition of the synchronization elements is unwanted there.

File: src/qdisp/Executive.cc [hunk 12]
```diff
@@ -72,14 +72,17 @@
...
 extern XrdSsiProvider *XrdSsiProviderClient;
 
 namespace {
 
 LOG_LOGGER _log = LOG_GET("lsst.qserv.qdisp.Executive");
 
-std::string getErrorText(XrdSsiErrInfo & e) {
-    std::ostringstream os;
+string getErrorText(XrdSsiErrInfo & e) {
+    ostringstream os;
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/qdisp/Executive.cc:85]: I feel it's usually unhelpful chaff chewing into the 110 char line limit. ¯\_(ツ)_/¯

File: src/wbase/SendChannelShared.h [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+    ///
+    /// @return true if inLast is true and this is the last task to call this
+    ///              with inLast == true.
+    /// The calling Thread must hold 'streamMutex' before calling this.
+    bool transmitTaskLast(StreamGuard sLock, bool inLast);
+
+    /// streamMutex is used to protect _lastCount and messages that are sent
+    /// using SendChannelShared.
+    std::mutex streamMutex;
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/wbase/SendChannelShared.h:128]: I have my doubts as well. So I'll look closer. The original code required it to keep the buffers in order. Code has moved and changed since then, so I don't think it needs to be outside, but there may still be cases where data is added and the mutex would be needed to keep the data in order. 

@iagaponenko is fond of passing lock_guards into functions that require mutex locks as an indication that the lock is held. That has shortcomings, such as there's no proof that the lock_guard has the correct mutex. lock_guards and mutex are not genric, if  there's one case where you need a unique_lock, all the locks have to be unique_lock or you can't pass it as a parameter. If a normal mutex doesn't work, you can have an issue. If the parent class requires lock_guard, and child class needs a recursive mutex and an occasional unique_lock, it wont be pretty.
Basically, it's just an indication, not proof, and it locks you into one type of lock+mutex. If you need a try_lock or something odd, it doesn't cope well.

[the remark this answers is not in the document]
Comment (andy-slac) [reply src/wbase/SendChannelShared.h:128]: I think the patter that Igor (and me) uses applies to private methods, in your own class you certainly have more control over what kind of lock to use. If you expose a mutex as public then this pattern is certainly questionable.

File: src/proto/worker.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgates108) [reply src/proto/worker.proto:102]: I've removed several required fields and put in checks to make sure that they are in the received messages. Most were fields added in the tickets related to this ticket, but some were older. It still seems like a good time for house cleaning.

File: src/wcontrol/Foreman.cc [hunk 14]
```diff
@@ -95,17 +97,17 @@ Foreman::~Foreman() {
     _pool->shutdownPool();
 }
 
-/// Put the task on the scheduler to be run later.
-void Foreman::processTask(std::shared_ptr<wbase::Task> const& task) {
 
+void Foreman::_setRunFunc(shared_ptr<wbase::Task> const& task) {
     auto func = [this, task](util::CmdData*){
```
[the remark this answers is not in the document]
Comment (jgates108) [reply src/wcontrol/Foreman.cc:102]: What is your basis for this? The only purpose of this function is to attach this lambda function to the task.

[the remark this answers is not in the document]
Comment (andy-slac) [reply src/wcontrol/Foreman.cc:102]: My point is that this lambda does not have to be a lambda, as it is a reasonably complex function. Any non-trivial lambda longer than couple of lines of code should be made a regular named function (and have a unit test written for it).

**Comment by iagaponenko** [comment]
@jgates108 based on the last round of extensive testings, this PR looks good.
